src/UnitTests/AttributeBasedMaps.cs

```diff
@@ -606,5 +607,221 @@ public void Should_validate_successfully()
+                    set { _value = value; }
+                }
+
+                public Dest(int addend)
+                {
+                    _addend = addend;
+                }
+
+                public Dest() : this(0)
+                {
+                }
+            }
+
+            protected override MapperConfiguration Configuration { get; } = new MapperConfiguration(cfg =>
+            {
+                cfg.CreateMissingTypeMaps = false;
+                cfg.AddMaps(typeof(When_specifying_to_construct_using_service_locator_via_attribute));
+                cfg.ConstructServicesUsing(t => new Dest(10));
+            });
+
+            [Fact]
+            public void Should_map_with_the_custom_constructor()
+            {
+                var source = new Source { Value = 6 };
+                var dest = Mapper.Map<Dest>(source);
+                dest.Value.ShouldBe(16);
+            }
+        }
+
+        public class When_specifying_max_depth_via_attribute : NonValidatingSpecBase
+        {
+            public class Source
+            {
+                public int Level { get; set; }
+                public IList<Source> Children { get; set; }
+                public Source Parent { get; set; }
+
+                public Source(int level)
+                {
+                    Children = new List<Source>();
+                    Level = level;
+                }
+
+                public void AddChild(Source child)
+                {
+                    Children.Add(child);
+                    child.Parent = this;
+                }
+            }
+
+            [AutoMap(typeof(Source), MaxDepth = 2)]
+            public class Dest
+            {
+                public int Level { get; set; }
+                public IList<Dest> Children { get; set; }
+                public Dest Parent { get; set; }
+            }
+
+            protected override MapperConfiguration Configuration { get; } = new MapperConfiguration(cfg =>
+            {
+                cfg.CreateMissingTypeMaps = false;
+                cfg.AddMaps(typeof(When_specifying_max_depth_via_attribute));
+            });
+
+            [Fact]
+            public void Third_level_children_are_null_with_max_depth_2()
+            {
+                var source = new Source(1);
+                source.AddChild(new Source(2));
+                source.AddChild(new Source(3));
+                source.Children[0].AddChild(new Source(4));
+                source.Children[1].AddChild(new Source(5));
+
+                var dest = Mapper.Map<Dest>(source);
+                dest.Level.ShouldBe(1);
+                dest.Children[0].Level.ShouldBe(2);
+                dest.Children[0].Children.ShouldAllBe(d => d == null);
+                dest.Children[1].Level.ShouldBe(3);
+                dest.Children[1].Children.ShouldAllBe(d => d == null);
+            }
+        }
+
+        public class When_specifying_to_preserve_references_via_attribute : NonValidatingSpecBase
```

Comment:
This test passes regardless of the attribute setting. Because of the built in logic to prevent SO. Try a non recursive mapping.

Reply:
I changed the test!